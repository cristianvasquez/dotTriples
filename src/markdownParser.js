import remarkFrontmatter from 'remark-frontmatter'
import remarkParse from 'remark-parse'
import { unified } from 'unified'

function createMarkdownParser () {
  const parser = unified().use(remarkParse).use(remarkFrontmatter)
  return {
    parse: async (string) => {
      return parser.parse(string)
    }
  }
}

export {createMarkdownParser}
