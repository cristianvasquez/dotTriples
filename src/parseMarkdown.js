import { YozoraParser } from '@yozora/parser'

const parser = new YozoraParser()
const ast = parser.parse('source content')

console.log(JSON.stringify(ast,null,2))