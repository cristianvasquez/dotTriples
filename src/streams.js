import { PassThrough } from 'stream'

function printJsons () {
  return createOutputStream({
    forEach: (chunk => {
      console.log(JSON.stringify(chunk, null, 2))
    }),
  })
}

function printDatasets () {
  return createOutputStream({
    forEach: (chunk => {
      console.log(chunk.toString())
    }),
  })
}

function printDatasetCounts () {
  return new PassThrough({
    objectMode: true, write (dataset, encoding, callback) {
      this.count = this.count ? this.count + dataset.size : dataset.size
      console.log(`${this.count} quads`)
      this.push(dataset)
      callback()
    },
  })
}

function createOutputStream ({ forEach } = { forEach: x => x }) {
  return new PassThrough({
    objectMode: true, write (object, encoding, callback) {
      forEach(object)
      this.push(object)
      callback()
    },
  })
}

export {
  printJsons,
  printDatasets,
  printDatasetCounts,
  createOutputStream,
}
