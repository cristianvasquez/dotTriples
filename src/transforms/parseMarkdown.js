import { Transform } from 'stream'

const ofInterest = (root) => (root.type === 'link' || root.type ===
  'wikiLink' || root.type === 'image')

class ParseMarkdown extends Transform {
  constructor ({ markdownParser }, header, opts) {
    super({ ...opts, objectMode: true })
    this.markdownParser = markdownParser
    this.header = header
  }

  collectChilds (parent, data = []) {
    if (parent.children) {
      return parent.children.reduce(
        (accumulator, currentValue) => {
          if (ofInterest(currentValue)) {
            return [currentValue, ...accumulator]
          } else {
            return this.collectChilds(currentValue, accumulator)
          }
        }, data)
    } else {
      return ofInterest(parent) ? [parent, ...data] : data
    }
  }

  async _transform (chunk, encoding, callback) {
    const fullText = chunk.toString()
    const markdown = await this.markdownParser.parse(fullText)
    for (const child of markdown.children) {

      const {
        type,
        value,
        position: {
          start: { offset: startOffset },
          end: { offset: endOffset },
        },
      } = child
      const text = fullText.substring(startOffset, endOffset)
      const header = { ...this.header, startOffset, endOffset }

      const links = this.collectChilds(child).map(node => {

        // sometimes the parser doesn't report a position. Weird.
        // @TODO report bug. example:'mailto:asset+block@lectureslides_chap3NOphantom.pdf'
        const linkText = node.position ? fullText.substring(
          node.position.start.offset,
          node.position.end.offset) : node.url
        const result = {
          ...node,
          text: linkText,
        }
        delete result.children
        delete result.position
        return result
      })

      this.push({
        header,
        type,
        text,
        value,
        links,
      })
    }
    callback()
  }
}

export { ParseMarkdown }
