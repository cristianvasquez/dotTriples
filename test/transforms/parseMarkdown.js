import { expect } from 'expect'

import toMatchSnapshot from 'expect-mocha-snapshot'
import { createReadStream } from 'fs'
import { resolve } from 'path'
import { createMarkdownParser } from '../../src/markdown/markdownParser.js'
import { ParseMarkdown } from '../../src/transforms/parseMarkdown.js'
import {collect} from '../support/streams.js'

expect.extend({ toMatchSnapshot })

const files = ['links.md']
const basePath = 'test/transforms/support'
const markdownParser = createMarkdownParser()

describe('[pipelines.js]', async function () {

  files.forEach((current) => {
    it(`"${current}"`, async function () {
      const filePath = resolve(basePath, current)
      const stream = createReadStream(filePath).
        pipe(new ParseMarkdown({ markdownParser }, {}, {}))
      const data = await collect(stream)
      expect(data).toMatchSnapshot(this)
    })
  })

})
