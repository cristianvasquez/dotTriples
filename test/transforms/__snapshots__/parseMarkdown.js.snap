// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[pipelines.js] "links.md" 1`] = `
Array [
  Object {
    "header": Object {
      "endOffset": 17,
      "startOffset": 0,
    },
    "links": Array [],
    "text": "## Markdown links",
    "type": "heading",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 56,
      "startOffset": 19,
    },
    "links": Array [
      Object {
        "text": "[a link to a url](http://example.org)",
        "title": null,
        "type": "link",
        "url": "http://example.org",
      },
    ],
    "text": "[a link to a url](http://example.org)",
    "type": "paragraph",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 92,
      "startOffset": 58,
    },
    "links": Array [
      Object {
        "text": "[a link to an entity](./entity.md)",
        "title": null,
        "type": "link",
        "url": "./entity.md",
      },
    ],
    "text": "[a link to an entity](./entity.md)",
    "type": "paragraph",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 112,
      "startOffset": 94,
    },
    "links": Array [],
    "text": "## Wikimedia links",
    "type": "heading",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 137,
      "startOffset": 114,
    },
    "links": Array [
      Object {
        "data": Object {
          "alias": "Entity",
          "exists": false,
          "hChildren": Array [
            Object {
              "type": "text",
              "value": "Entity",
            },
          ],
          "hName": "a",
          "hProperties": Object {
            "className": "internal new",
            "href": "#/page/entity",
          },
          "permalink": "entity",
        },
        "text": "[[Entity]]",
        "type": "wikiLink",
        "value": "Entity",
      },
    ],
    "text": "A link to an [[Entity]]",
    "type": "paragraph",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 154,
      "startOffset": 139,
    },
    "links": Array [],
    "text": "## Markdown url",
    "type": "heading",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 191,
      "startOffset": 156,
    },
    "links": Array [
      Object {
        "text": "<http://web2.org>",
        "title": null,
        "type": "link",
        "url": "http://web2.org",
      },
      Object {
        "text": "<http://web1.org>",
        "title": null,
        "type": "link",
        "url": "http://web1.org",
      },
    ],
    "text": "<http://web1.org>
<http://web2.org>",
    "type": "paragraph",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 206,
      "startOffset": 193,
    },
    "links": Array [],
    "text": "## Simple url",
    "type": "heading",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 239,
      "startOffset": 208,
    },
    "links": Array [
      Object {
        "text": "http://web4.org",
        "title": null,
        "type": "link",
        "url": "http://web4.org",
      },
      Object {
        "text": "http://web3.org",
        "title": null,
        "type": "link",
        "url": "http://web3.org",
      },
    ],
    "text": "http://web3.org
http://web4.org",
    "type": "paragraph",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 249,
      "startOffset": 241,
    },
    "links": Array [],
    "text": "## Lists",
    "type": "heading",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 275,
      "startOffset": 251,
    },
    "links": Array [
      Object {
        "data": Object {
          "alias": "Entity",
          "exists": false,
          "hChildren": Array [
            Object {
              "type": "text",
              "value": "Entity",
            },
          ],
          "hName": "a",
          "hProperties": Object {
            "className": "internal new",
            "href": "#/page/entity",
          },
          "permalink": "entity",
        },
        "text": "[[Entity]]",
        "type": "wikiLink",
        "value": "Entity",
      },
    ],
    "text": "- Hello
    - [[Entity]]",
    "type": "list",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 288,
      "startOffset": 277,
    },
    "links": Array [],
    "text": "## An image",
    "type": "heading",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 323,
      "startOffset": 290,
    },
    "links": Array [
      Object {
        "text": "[Lovely image](../folder/img.png)",
        "title": null,
        "type": "link",
        "url": "../folder/img.png",
      },
    ],
    "text": "[Lovely image](../folder/img.png)",
    "type": "paragraph",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 347,
      "startOffset": 326,
    },
    "links": Array [],
    "text": "## With transclusion ",
    "type": "heading",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 383,
      "startOffset": 349,
    },
    "links": Array [
      Object {
        "alt": "Lovely image",
        "text": "![Lovely image](../folder/img.png)",
        "title": null,
        "type": "image",
        "url": "../folder/img.png",
      },
    ],
    "text": "![Lovely image](../folder/img.png)",
    "type": "paragraph",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 423,
      "startOffset": 385,
    },
    "links": Array [
      Object {
        "alt": "a link to a url",
        "text": "![a link to a url](http://example.org)",
        "title": null,
        "type": "image",
        "url": "http://example.org",
      },
    ],
    "text": "![a link to a url](http://example.org)",
    "type": "paragraph",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 456,
      "startOffset": 425,
    },
    "links": Array [],
    "text": "### (currently not implemented)",
    "type": "heading",
    "value": undefined,
  },
  Object {
    "header": Object {
      "endOffset": 480,
      "startOffset": 458,
    },
    "links": Array [],
    "text": "![[TranscludedEntity]]",
    "type": "paragraph",
    "value": undefined,
  },
]
`;
